Save unit test results as JUnit XML for Shippable

ixdy opened the pull request:
Shippable will parse JUnit-style XML test results if they exist (http://docs.shippable.com/yml_reference/#test-results).

A little bit of hackery is needed to get our unit tests infrastructure to produce a JUnit-style test report, but this seems to work now.

This may help #10300. It will also be useful should we end up running unit tests on Jenkins.

ixdy committed:
Add option to save unit test results as JUnit XML for Shippable

ixdy commented:
(still testing Shippable - don't review/merge yet)

ixdy commented:
Seems like this actually works. @zmerlynn want to take a look at some terrifying bash?

zmerlynn commented:
Holy crap yes! I was just trying to poke at how to possibly go about this, because I'm so tired of the bullshit `go test` output and how long it takes it to look through versus our Jenkins report. Reviewing.

zmerlynn commented on hack/test-go.sh:
```diff
@@ -117,6 +125,33 @@ if [[ ${#testcases[@]} -eq 0 ]]; then
+    echo ""
+    return
+  fi
+  mkdir -p ${KUBE_JUNIT_REPORT_DIR}
```
nit: quote `"${KUBE_JUNIT_REPORT_DIR}"`

zmerlynn commented on hack/test-go.sh:
```diff
@@ -117,6 +125,33 @@ if [[ ${#testcases[@]} -eq 0 ]]; then
+    return
+  fi
+  mkdir -p ${KUBE_JUNIT_REPORT_DIR}
+  echo ${KUBE_JUNIT_REPORT_DIR}/junit_${KUBE_API_VERSION}_$(kube::util::sortable_date)
```
Ditto (might as well just quote the line).

zmerlynn commented on hack/test-go.sh:
```diff
@@ -117,6 +125,33 @@ if [[ ${#testcases[@]} -eq 0 ]]; then
+}
+
+produceJUnitXMLReport() {
+  local junit_filename_prefix=${1:-""}
```
`local -r`

zmerlynn commented on hack/test-go.sh:
```diff
@@ -117,6 +125,33 @@ if [[ ${#testcases[@]} -eq 0 ]]; then
+    return
+  fi
+
+  local test_stdout_filenames=$(ls ${junit_filename_prefix}*.stdout)
```
```
local test_stdout_filenames
test_stdout_filenames=$(ls ${junit_filename_prefix}*.stdout)
```

zmerlynn replied:
(Otherwise errexit is ignored.)

zmerlynn commented on hack/test-go.sh:
```diff
@@ -152,14 +187,18 @@ runTests() {
     fi
   fi
 
+  local junit_filename_prefix=$(junitFilenamePrefix)
```
Ditto here, local split.

zmerlynn commented on hack/test-go.sh:
```diff
@@ -152,14 +187,18 @@ runTests() {
-    return 0
+     "${testargs[@]:+${testargs[@]}}" \
+     | tee ${junit_filename_prefix:+"${junit_filename_prefix}.stdout"} && rc=$? || rc=$?
+    produceJUnitXMLReport ${junit_filename_prefix}
```
Nit: `"${junit_filename_prefix}"`

zmerlynn commented on hack/test-go.sh:
```diff
@@ -152,14 +187,18 @@ runTests() {
-     "${testargs[@]:+${testargs[@]}}"
-    return 0
+     "${testargs[@]:+${testargs[@]}}" \
+     | tee ${junit_filename_prefix:+"${junit_filename_prefix}.stdout"} && rc=$? || rc=$?
```
Nit: Just quote the whole ternary, `"${junit_filename_prefix:+${junit_filename_prefix}.stdout}"

ixdy replied:
if I do this, then when `$junit_filename_prefix` is empty, it tries to run `tee ""` (I think), which seems to fail:
```console
tee: : No such file or directory
```

zmerlynn replied:
Got it.

zmerlynn commented on hack/test-go.sh:
```diff
@@ -173,15 +212,23 @@ runTests() {
   printf "%s\n" "${@}" | xargs -I{} -n1 -P${KUBE_COVERPROCS} \
-      go test "${goflags[@]:+${goflags[@]}}" \
+    sh -c "set -o pipefail; _pkg=\"{}\"; _pkg_out=\${_pkg//\//_}; \
+        go test ${goflags[@]:+${goflags[@]}} \
```
nit: quote `"${goflags[@]:+${goflags[@]}}"` (as the left)

ixdy replied:
same as noted for `$testargs` below.

zmerlynn replied:
Yeah, okay.

zmerlynn commented on hack/test-go.sh:
```diff
@@ -173,15 +212,23 @@ runTests() {
+          -cover -covermode=\"${KUBE_COVERMODE}\" \
+          -coverprofile=\"${cover_report_dir}/\${_pkg}/${cover_profile}\" \
+          \"${KUBE_GO_PACKAGE}/\${_pkg}\" \
+          ${testargs[@]:+${testargs[@]}} \
```
nit: quote `"${testargs[@]:+${testargs[@]}}"` (as the left)

ixdy replied:
this was intentional. without it, `go test` fails when testargs isn't set with
```
can't load package: package .: no buildable Go source files in /usr/local/google/home/jgrafton/code/kubernetes/src/github.com/kubernetes/kubernetes
can't load package: package .: no buildable Go source files in /usr/local/google/home/jgrafton/code/kubernetes/src/github.com/kubernetes/kubernetes
```

zmerlynn replied:
I love shell.

zmerlynn commented on hack/test-go.sh:
```diff
@@ -173,15 +212,23 @@ runTests() {
+          -coverprofile=\"${cover_report_dir}/\${_pkg}/${cover_profile}\" \
+          \"${KUBE_GO_PACKAGE}/\${_pkg}\" \
+          ${testargs[@]:+${testargs[@]}} \
+        | tee ${junit_filename_prefix:+\"$junit_filename_prefix-\$_pkg_out.stdout\"}" \
```
nit: quote the ternary

zmerlynn commented on hack/test-go.sh:
```diff
@@ -173,15 +212,23 @@ runTests() {
+        | tee ${junit_filename_prefix:+\"$junit_filename_prefix-\$_pkg_out.stdout\"}" \
+      && test_result=$? || test_result=$?
+
+  produceJUnitXMLReport ${junit_filename_prefix}
```
and quote this too, while we're at it. :)

zmerlynn commented on hack/test-go.sh:
```diff
@@ -152,14 +187,18 @@ runTests() {
+     "${testargs[@]:+${testargs[@]}}" \
+     | tee ${junit_filename_prefix:+"${junit_filename_prefix}.stdout"} && rc=$? || rc=$?
+    produceJUnitXMLReport ${junit_filename_prefix}
+    return $rc
```
`${rc}` for consistency (the file is strongly `${var}` consistent, which is also style guide preferred)

zmerlynn commented on hack/test-go.sh:
```diff
@@ -201,6 +248,8 @@ runTests() {
   go tool cover -html="${COMBINED_COVER_PROFILE}" -o="${coverage_html_file}"
   kube::log::status "Combined coverage report: ${coverage_html_file}"
+
+  return $test_result
```
`${test_result}` for consistency.

zmerlynn commented:
Bunch of nits, LGTM otherwise

ixdy commented:
Addressed all of your nits (I think). I also added an intentionally failing test, just to see what Shippable does when that happens. After it's done I'll revert that commit (and squash).

zmerlynn commented:
Yeah, LGTM. Will give you the label when we see what it looks like and the test is gone.

ixdy commented:
Example test failure on Shippable: https://app.shippable.com/builds/55dce36a36590c0b00d46172

Squashed commits and reverted the failing test.

ixdy commented:
welp, DO NOT MERGE this yet.

From https://app.shippable.com/builds/55dcfd7d49da060b00aec71e:
```
sh: 1: set: Illegal option -o pipefail
```

ixdy commented:
OK, replaced `sh -c` with `bash -c` - should be safe on Shippable now.

zmerlynn commented:
LGTM
